src/services/ContributingDeveloperAuditService/scm/GitHub/api/GitHubApiClient.ts

```diff
@@ -0,0 +1,218 @@
+import axios, { AxiosInstance, AxiosResponse } from "axios";
+import { soosLogger } from "../../../../../logging/SOOSLogger";
+import { sleep } from "../../../../../utilities";
+
+interface IHttpRequestParameters {
+  baseUri: string;
+  githubPAT: string;
+}
+
+interface IHttpClientParameters extends IHttpRequestParameters {
+  apiClientName: string;
+}
+
+export interface GitHubOrganization {
+  login: string;
+}
+
+export interface GitHubRepository {
+  id: number;
+  name: string;
+  full_name: string;
+  private: boolean;
+  owner: GitHubOrganization;
+}
+
+export interface ContributingDeveloper {
+  username: string;
+  repositories: ContributingDeveloperRepositories[];
+}
+
+interface ContributingDeveloperRepositories {
+  id: number;
+  name: string;
+  lastCommit: string;
+  isPrivate: boolean;
+}
+
+export interface Commits {
+  commit: {
+    author: Author;
+  };
+}
+
+export interface Author {
+  name: string;
+  email: string;
+  date: string;
+}
+
+export const threeMonthsDate = (() => {
+  const d = new Date();
+  d.setDate(d.getDate() - 90);
+  return `${d.getUTCFullYear()}-${
+    d.getMonth() + 1
+  }-${d.getUTCDate()}T${d.getHours()}:${d.getMinutes()}:${d.getSeconds()}Z`;
+})();
+
+class GitHubApiClient {
+  private readonly client: AxiosInstance;
+
+  constructor(githubPAT: string, baseUri: string = "https://api.github.com") {
+    this.client = GitHubApiClient.createHttpClient({
+      baseUri,
+      githubPAT,
+      apiClientName: "GitHub API",
+    });
+  }
+
+  private static createHttpClient({ baseUri, githubPAT, apiClientName }: IHttpClientParameters) {
+    const client = axios.create({
+      baseURL: baseUri,
+      headers: {
+        accept: "application/vnd.github+json",
+        Authorization: `Bearer ${githubPAT}`,
+      },
+    });
+
+    client.interceptors.request.use(
+      (request) => {
+        if (request.data) {
+          soosLogger.verboseDebug(
+            apiClientName,
+            `Request URL: ${request.method?.toLocaleUpperCase()} ${request.url}`,
+          );
+          if (request.params) {
+            soosLogger.verboseDebug(
+              apiClientName,
+              `Request Params: ${JSON.stringify(request.params)}`,
+            );
+          }
+          soosLogger.verboseDebug(apiClientName, `Request Body: ${JSON.stringify(request.data)}`);
+        }
+        return request;
+      },
+      (rejectedRequest) => {
+        return Promise.reject(rejectedRequest);
+      },
+    );
+
+    client.interceptors.response.use(
+      async (response) => {
+        soosLogger.verboseDebug(apiClientName, `Response Body: ${JSON.stringify(response.data)}`);
```

Comment:
We might want to just remove this line, it creates a bunch of noise in the console since there are a bunch of requests and responses are really lengthy

Reply:
another option is to make the interceptor optional so we can use it during development, but it won't be used when the script is run

Reply:
or make it a flag on the interceptor function

Reply:
I tried it - it isn't worth it - not enough reuse because of the response handling